usb/usb_host_hid/hid_host.c

```diff
@@ -1254,13 +1277,29 @@ esp_err_t hid_host_device_close(hid_host_device_handle_t hid_dev_handle)
     return ESP_OK;
 }
 
-esp_err_t hid_host_handle_events(uint32_t timeout_ms)
+esp_err_t hid_host_handle_events(uint32_t timeout)
 {
     HID_RETURN_ON_FALSE(s_hid_driver,
                         ESP_ERR_INVALID_STATE,
                         "HID Driver is not installed");
 
-    return usb_host_client_handle_events(s_hid_driver->client_handle, timeout_ms);
+    if (!s_hid_driver->ext_task_config.active) {
+        return ESP_ERR_NOT_SUPPORTED;
+    }
+
+    esp_err_t ret = usb_host_client_handle_events(s_hid_driver->client_handle, timeout);
+
+    if (s_hid_driver->end_client_event_handling) {
+        HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
+                             "HID Client deregister failure");
+
+        xSemaphoreGive(s_hid_driver->all_events_handled);
+        // wait for hid_host_uninstall completion
+        s_hid_driver->ext_task_config.external_task_hdl = xTaskGetCurrentTaskHandle();
```

Comment:
`external_task_hdl` is set in function hid_host_handle_events() in one thread and then referred in function hid_host_uninstall() in another thread. These two threads can race against each other. I think, its better to set `external_task_hdl` at the very start if function hid_host_handle_events(). It will reduce the possibility race condition.

Something like this:

```
esp_err_t hid_host_uninstall(void)
{
    ***
    if (s_hid_driver->ext_task_config.active &&
        !s_hid_driver->ext_task_config.external_task_hdl) {
            ESP_LOGE(TAG, "HID Host external event handling has not been called");
            // Deregister client by ourselves and release semaphore
            HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                                 "HID Client deregister failure");
            xSemaphoreGive(s_hid_driver->all_events_handled);
        }
    }

    xSemaphoreTake(s_hid_driver->all_events_handled, portMAX_DELAY);
    vSemaphoreDelete(s_hid_driver->all_events_handled);
    s_hid_driver->ext_task_config.external_task_hdl = NULL;
    s_hid_driver->ext_task_config.active = false;
    ***
}

esp_err_t hid_host_handle_events(uint32_t timeout)
{
    HID_RETURN_ON_FALSE(s_hid_driver,
                        ESP_ERR_INVALID_STATE,
                        "HID Driver is not installed");

    if (!s_hid_driver->ext_task_config.active) {
        return ESP_ERR_NOT_SUPPORTED;
    }
    s_hid_driver->ext_task_config.external_task_hdl = xTaskGetCurrentTaskHandle();

    esp_err_t ret = usb_host_client_handle_events(s_hid_driver->client_handle, timeout);

    if (s_hid_driver->end_client_event_handling) {
        HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                             "HID Client deregister failure");

        xSemaphoreGive(s_hid_driver->all_events_handled);
    }

    return ret;
}
```

Reply:
Hey @esp-saurabhbansal, 
Thanks for the review.

Probably, I don't get the point. Let me try to explain my logic, probably after that you can just point the moment when I am not right or has a logic misleading. 

1. When application logic calls the `hid_host_handle_events()`, we spend some time in `usb_host_client_handle_events()`. To make it compatible with power management features such as tickless idle and automatic light sleep we need to provide some mechanism with a possibility to use maximum delay `portMAX_DELAY` to wait next event and do not poll peripheral every N times per minute. This can be achieved when we call `hid_host_handle_events(portMAX_DELAY);` from the external polling task;
2. In this case, the host will wait another client event. If there is no event, the logic will return from this function after `usb_host_client_unblock()` during the `hid_host_uninstall()`;
3. Before client unblock, we change the state of internal flag `s_hid_driver->end_client_event_handling = true;` during uninstall;
4. When logic returns from `usb_host_client_handle_events()` in `hid_host_handle_events()` fulfilled in external task,  the  flag end_client_event_handling is true and we unblock the client, release semaphore and save the Task handle and suspend the external Task (here is another question, if there is more logic except the client events polling in this task, so we can halt this logic as well, which is not so good);

...

I see. 
Do you mean, that we can Resume the external Task before it (external Task) release semaphore and will go to Suspend, right?

Reply:
Lets call the thread as threadA which calls hid_host_uninstall() and threadB which calls hid_host_handle_events()

After Point 3, in case threadA keeps on executing and threadB does not get CPU, threadA will check for the if conditions and it finds external_task_hdl as NULL and it execute this block. It will call usb_host_client_deregister() -> xSemaphoreGive() -> ....
```
        if (NULL == s_hid_driver->ext_task_config.external_task_hdl) {
            ESP_LOGE(TAG, "HID Host external event handling has not been called");
            // Deregister client by ourselves and release semaphore
            HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                                 "HID Client deregister failure");
            xSemaphoreGive(s_hid_driver->all_events_handled);
        }
```


Next, threadB gets the CPU and starts executing. After finding s_hid_driver->end_client_event_handling as true, it will start executing this block:
```
    if (s_hid_driver->end_client_event_handling) {
        HID_RETURN_ON_ERROR( usb_host_client_deregister(s_hid_driver->client_handle),
                             "HID Client deregister failure");

        xSemaphoreGive(s_hid_driver->all_events_handled);
        // wait for hid_host_uninstall completion
        s_hid_driver->ext_task_config.external_task_hdl = xTaskGetCurrentTaskHandle();
        vTaskSuspend(NULL); // Task will be resumed from hid_host_uninstall()
    }
```

This threaB will again call usb_host_client_deregister() -> xSemaphoreGive() -> sets external_task_hdl -> vTaskSuspend itself.

This flow happens because threadA find the value of external_task_hdl as NULL.
So, my sugession was to set external_task_hdl with xTaskGetCurrentTaskHandle(); before calling usb_host_client_handle_events().

I can be unclear now also. We can have a call to discuss.